experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts

```diff
@@ -172,4 +173,100 @@ describe("Fuzz - Targeted", () => {
 			emitter,
 		});
 	});
+
+	const undoRedoWeights: Partial<EditGeneratorOpWeights> = {
+		insert:1,
+		delete:1,
+		start:0,
+		commit: 0
+	};
+
+	describe.only("Inorder undo/redo matches the initial/final state", () => {
+		const generatorFactory = (): AsyncGenerator<TreeOperation, UndoRedoFuzzTestState> =>
+			takeAsync(opsPerRun, makeOpGenerator(undoRedoWeights));
+
+		const model: DDSFuzzModel<
+			SharedTreeTestFactory,
+			Operation,
+			DDSFuzzTestState<SharedTreeTestFactory>
+		> = {
+			workloadName: "SharedTree",
+			factory: new SharedTreeTestFactory(onCreate),
+			generatorFactory,
+			reducer: fuzzReducer,
+			validateConsistency: () => {},
+		};
+		const emitter = new TypedEventEmitter<DDSFuzzHarnessEvents>();
+		emitter.on("testStart", (initialState: UndoRedoFuzzTestState) => {
+			initialState.initialTreeState = toJsonableTree(initialState.clients[0].channel);
+			initialState.firstAnchors = []
+			// creates an initial anchor for each tree
+			for (const client of initialState.clients){
+				initialState.firstAnchors.push(getFirstAnchor(client.channel))
+			}
+		});
+		emitter.on("testEnd", (finalState: UndoRedoFuzzTestState) => {
+			const clients = finalState.clients
+
+			const finalTreeStates = []
+			// undo all of the changes and validate against initialTreeState for each tree
+			for (const [i, client] of clients.entries()){
+				const tree = client.channel;
+
+				// save final tree states to validate redo later
+				finalTreeStates.push(toJsonableTree(tree))
+
+				// call undo() until tree no more edits left
+				for (let j = 0; j<opsPerRun; j++){
+					tree.undo()
+				}
```

Comment:
@yann-achard-MS I am currently just using a loop and calling the `tree.undo()` method `opsPerRun` number of times. This is a more "lazy" way of doing this, as we only produce at max this many number of ops. But would there be a better way to do this? One way would be to check the `UndoRedoManager` in the tree's branch to check if the head is undefined, but we do not have access to this as it is not exposed.

Reply:
SharedTree is supposed to be exposing an undo stack size ([work item](https://dev.azure.com/fluidframework/internal/_workitems/edit/3625)) so I'm reticent to invest too much in something more refined than what you have here.

Reply:
seems fine to me, assuming it's part of the spec that undo no-ops when it reaches an empty stack.